### ISSUE TITLE
Embedded code cell for flexible programatic value manipulation

### ISSUE THREAD

**Issue by nthiery** [sagemath/sage-explorer#29]
# Context

Manipulating objects by UI gestures as in the explorer, or by traditional command line both have their merits. So we want the user to be able to easily switch from one to the other. When we do:

```
w = explore(o)
w
```
the user can already access and change the currently displayed value by going to any code cell outside of the explorer and using `w.value`. For occasional manipulations, that's ok, but it becomes cumbersome to switch back and forth. Also the user must have named the explorer; so this does not work with just `explore(o)`.

# Proposal

Embed in the explorer a code cell. Code entered there is evaluated in the usual global namespace of other cells, with a few local values:
- `_` refers to the current value (shorthand for `w.value`)
- `Hist` (or History?) refers to the history of the explorer (shorthand for `w.history`)
- `__explorer__` refers to the explorer
If the code evaluates to a non None value, then this value becomes the new explored value.

A tooltip (or what's this button?) would explain the above to the user.

# Scenario

The user runs an explorer:
```
   explore(o)
```
and explores around. At some point, she finds an interesting value that she wants to name it for later use. She enters in the code cell:
```
   t = _
```
Later on, she stumbles on an object and would like to apply some operation that is cumbersome to express with a method call. To do some arithmetic, e.g. with previous values, named or not, she can do
```
_ + 3 * t + 2*Hist[2]
```
Similarly if she could do indexed operations on a container such as:
```
_[3]
```
```
_[5] += 3
```

# Implementation details

The evaluation of the code could look like:
```
    g = globals() # the name space used by the usual Jupyter cells
    l = {"_":self.value, "__explorer__":self, "Hist": self.history}
    local_names = l.keys()
    code = compile(code, '<string>', 'single')
    result = eval(code, g, l)
    if result is None: # the code may have triggered some assignments
        for name, value in l.items():
            if name not in local_names:
            g[name] = value
    else:
         self.value = result
```
Note: we need to use `compile` to support assignments, as I learned from this [tutorial](http://www.python-simple.com/python-langage/evaluation.php).
# Nice to have

Syntax highlighting, code completion, ... as in a usual Jupyter cell

**Comment by zerline** [comment]
With https://github.com/sagemath/sage-explorer/commit/c12dbb2a6f8e8caa78d26c01c02d9dd7bb0e0edb,
we now have a basic code cell that evaluates expressions, including:

- _ and __explorer__._history.initial_name (name alrady given by the user to the object, if any)

- Hist

- __explorer__

**Comment by nthiery** [comment]
Confirmed! With master, commit 0d18e10d1d5107506617ba19bc0e6cceb72dd1f6, the evaluation seems to not happen in Sage's global name space. E.g. inputing ZZ raises an error "name ZZ is undefined".

**Comment by zerline** [comment]
Closed in favor of #49
